documentation/src/main/asciidoc/reference/search-dsl-predicate.asciidoc

```diff
@@ -784,11 +784,143 @@ but can be <<search-dsl-predicate-common-boost,boosted>>,
 either on a per-field basis with a call to `.boost(...)` just after `.field(...)`/`.fields(...)`
 or for the whole predicate with a call to `.boost(...)` after `.matchingAny(...)` or `.matchingAll(...)`.
 
+[[search-dsl-predicate-and]]
+== `and`: match all clauses
+
+The `and` predicate matches documents that match **all** of its inner predicates, called "clauses".
+
+Matching "and" clauses are taken into account during <<search-dsl-predicate-common-score,score>> computation.
+
+.Matching a document that matches all the multiple given predicates (~`AND` operator)
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/search/predicate/PredicateDslIT.java[tags=and]
+----
+<1> The hits *must* have a `title` field matching the text `robot`,
+independently of other clauses in the same boolean predicate.
+<2> The hits *must* have a `description` field matching the text `crime`,
+independently of other clauses in the same boolean predicate.
+<3> All returned hits will match *all* the clauses above:
+they will have a `title` field matching the text `robot`
+*and* they will have a `description` field matching the text `crime`.
+====
+
+[[search-dsl-predicate-and-lambda]]
+=== Adding clauses dynamically with the lambda syntax
+
+It is possible to define the `and` predicate inside a lambda expression.
+This is especially useful when clauses need to be added dynamically to the `and` predicate,
+for example based on user input.
+
+.Easily adding clauses dynamically using `.where(...)` and the lambda syntax
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/search/predicate/PredicateDslIT.java[tags=and-dynamicParameters-root]
+----
+<1> Get a custom object holding the search parameters provided by the user through a web form, for example.
+<2> Call `.where(Function)`.
+The function, implemented by a lambda expression, will receive a predicate factory, use it to build an `and` predicate,
+invoke the `with(Consumer)` method and return this predicate. The consumer, implemented by a lambda expression,
+will receive a clause collector for the `and` predicate as an argument, and will add clauses to that collector as necessary.
+<3> By default, a boolean predicate will match nothing if there is no clause.
+To match every document when there is no clause, add an `and` clause that matches everything.
+<4> Inside the lambda, the code is free to use any Java language constructs, such as `if` or `for`,
+to control the addition of clauses.
+In this case, we only add clauses if the relevant parameter was filled in by the user.
+====
+
+Another syntax relying on the method `with(...)` can be useful when the boolean predicate is not the root predicate,
+see <<example-and-or-dynamicParameters-with,this example>>.
+
```

Comment:
```suggestion
```

That's not "another" syntax, that's precisely the syntax described in the example above.